felix/bpf-gpl/conntrack_cleanup.c

```diff
@@ -24,229 +24,66 @@ struct ct_iter_ctx {
-			// link to the forward entry so we write a dummy key and use
-			// BPF_NOEXIST to avoid overwriting a real key.
-			struct calico_ct_key dummy_key = {};
-			long rc = cali_ccq_update_elem(ct_key, &dummy_key, BPF_NOEXIST);
-			if (rc) {
-				CALI_DEBUG("  EXPIRED: Reverse NAT entry, already in queue or queue full (rc=%d).", rc);
-			} else {
-				CALI_DEBUG("  EXPIRED: Reverse NAT entry, queued for deletion. (Forward NAT entry not yet seen.)");
-			}
+	} else if (isNatForward) {
+		// Its a NAT forward entry but the reverse entry is not present in the CT table.
+		// Set the lookup_key to the forward entry and do a lookup of the fwd entry.
+		// If the timestamp's match, delete it.
+		lookup_key = key;
+		actual_ct_value = cali_ct_lookup_elem(lookup_key);
+		// Reverse key is present, but it is already deleted from the CT table.
+		if (actual_ct_value && actual_ct_value->last_seen == value->last_seen) {
+			goto delete;
 		}
-		break;
+		return 0;
 	}
-
-	return 0;
-}
-
-// process_ccq_entry processes an entry in the "cleanup queue" map.  The map
-// is keyed on the reverse NAT entry's key, with the value storing the forward
-// entry's key (or a zero value if the forward entry is missing).
-static long process_ccq_entry(void *map, const void *key, void *value, void *ctx)
-{
-	// Map stores mapping from reverse key to forward key (if known).
-	const struct calico_ct_key *rev_key = key;
-	const struct calico_ct_key *fwd_key = value;
-	struct ct_iter_ctx *ictx = ctx;
-
-	// It might be a few ms since we queued this entry, recheck it to make sure
-	// it is still expired.
-	const struct calico_ct_value *rev_value = cali_ct_lookup_elem(rev_key);
-	if (rev_value) {
-		__u64 age = sub_age(ictx->now, rev_value->last_seen);
-		__u64 age_s = age/1000000000ull;
-#ifdef IPVER6
-		CALI_DEBUG("Re-checking: proto=%d [%pI6]:%d", rev_key->protocol, &rev_key->addr_a, rev_key->port_a);
-		CALI_DEBUG("  <->[%pI6]:%d age=%ds", &rev_key->addr_b, rev_key->port_b, age_s);
-#else
-		CALI_DEBUG("Re-checking: proto=%d %pI4:%d", rev_key->protocol, &rev_key->addr_a, rev_key->port_a);
-		CALI_DEBUG("  <->%pI4:%d age=%ds", &rev_key->addr_b, rev_key->port_b, age_s);
-#endif
-		__u64 max_age = calculate_max_age(rev_key, rev_value);
-		__u64 max_age_s = max_age/1000000000ull;
-		if (age < max_age) {
-			// Race with a packet, CT entry now live again.
-			CALI_DEBUG("  RESURRECTED: entry no longer expired (max_age=%d).", max_age_s);
-			goto out;
+delete:
+	// If the entry is Normal or Reverse, lookup_key points to the entry.
+	// Delete the entry and if successful, increment the counter.
+	// If the entry is forward, we have 2 cases,
+	// Rev_entry present - lookup_key points to the reverse entry and key points to the
+	// forward entry. Delete both the entries togethere.
+	// Rev_entry not present - lookup_key points to the forward entry.
+	// Delete it.
+	if(!cali_ct_delete_elem(lookup_key)) {
+		ictx->num_cleaned++;
+		if (isNatForward && !cali_ct_delete_elem(key)) {
```

Comment:
Not sure I grok why this is now nested; when there's a NAT pair, we should decide whether to delete both or neither and then either try to delete both or neither.

The first delete could fail due to LRU age out, for example.  (Pretty unlikely given we just looked it up but still...)